### ISSUE TITLE
this.userId is not always available onBeforeUpload

### ISSUE THREAD

**Issue by oeway** [VeliovGroup/Meteor-Files#207]
I am trying to validate the user in `onBeforeUpload`:
```js
 onBeforeUpload: function (file) {
    console.log(this.userId);
}
```
But on the server it outputs 3 lines for a single upload:
```
I20160830-12:46:55.587(2)? undefined
I20160830-12:46:55.820(2)? LSmRXxxkSs9RrPfN4
I20160830-12:46:55.832(2)? LSmRXxxkSs9RrPfN4
```
Two questions, one is why it calls three times? second is why the first time `this.userId` is not defined?
Thanks.

**Comment by dr-dimitru** [comment]
>  why it calls three times?

`onBeforeUpload` on server triggered for "first byte", each received chunk, and for "EOF"

> why the first time this.userId is not defined?

How do you initiate upload?

**Comment by oeway** [comment]
This is how I do the upload:
```
        var upload = Files.insert({
          file: file,
          transport: 'http',
          streams: 'dynamic',
          chunkSize: 'dynamic',
          meta: {
          }
        }, false);
        //setup callbacks
        //...
        upload.start();

```

**Comment by dr-dimitru** [comment]
Could you please post requests content (need to see headers)?
Can be obtained in network tab.

**Comment by dr-dimitru** [comment]
And please, make sure user is logged in when initiates an upload

**Comment by oeway** [comment]
For the first question, the behavior of running onBeforeUpload on each step is actually a little bit confusing to me, because I would think it should only execute once as it name suggest. So if I did a lot of DB operation in onBeforeUpload, it will actually slow down the server a lot, is there a way to execute only once?

**Comment by oeway** [comment]
There are three __upload post requests:
```
headers:

POST /cdn/storage/files/__upload HTTP/1.1
Host: cabat.imod.pasteur.fr:3000
Connection: keep-alive
Content-Length: 336
Pragma: no-cache
Cache-Control: no-cache
x-start: 1
Origin: http://cabat.imod.pasteur.fr:3000
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2743.82 Safari/537.36
x-mtok: Q72tzjS9zbmXQW4ih
Content-Type: application/json
Accept: */*
DNT: 1
Referer: http://cabat.imod.pasteur.fr:3000/widgetboard
Accept-Encoding: gzip, deflate
Accept-Language: en,zh;q=0.8,zh-CN;q=0.6,zh-TW;q=0.4
Cookie: __utma=219899544.213272128.1414491796.1445026431.1445347606.8; EMF_VISIT__MD_414=C2F04569-EC57-4774-A903-91EC1815; _ga=GA1.2.213272128.1414491796; __unam=4127071-1495649a3ce-65939ba9-116; meteor_login_token=eiWZxP_ln4c_--4S3VkqWgYLrPoygncjEsbt79YOwn4; x_mtok=Q72tzjS9zbmXQW4ih

payload:

{"file":{"size":31995,"type":"image/jpeg","name":"test.jpeg","meta":{"taskId":"kJjhWNeompFZ69spz","widgetId":"EejuZZ7munH67mJFt","tags":[],"owner":"LSmRXxxkSs9RrPfN4"},"ext":"jpeg","extension":"jpeg","extensionWithDot":".jpeg","mime":"image/jpeg","mime-type":"image/jpeg"},"fileId":"css36DL8hyPp7sybn","chunkSize":163840,"fileLength":1}
```

```
headers:

POST /cdn/storage/files/__upload HTTP/1.1
Host: cabat.imod.pasteur.fr:3000
Connection: keep-alive
Content-Length: 42660
Pragma: no-cache
Cache-Control: no-cache
Origin: http://cabat.imod.pasteur.fr:3000
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2743.82 Safari/537.36
x-mtok: Q72tzjS9zbmXQW4ih
content-type: text/plain
x-chunkid: 1
x-fileid: css36DL8hyPp7sybn
Accept: */*
DNT: 1
Referer: http://cabat.imod.pasteur.fr:3000/widgetboard
Accept-Encoding: gzip, deflate
Accept-Language: en,zh;q=0.8,zh-CN;q=0.6,zh-TW;q=0.4
Cookie: __utma=219899544.213272128.1414491796.1445026431.1445347606.8; EMF_VISIT__MD_414=C2F04569-EC57-4774-A903-91EC1815; _ga=GA1.2.213272128.1414491796; __unam=4127071-1495649a3ce-65939ba9-116; meteor_login_token=eiWZxP_ln4c_--4S3VkqWgYLrPoygncjEsbt79YOwn4; x_mtok=Q72tzjS9zbmXQW4ih

payload:
...

```

```
headers:
POST /cdn/storage/files/__upload HTTP/1.1
Host: cabat.imod.pasteur.fr:3000
Connection: keep-alive
Content-Length: 0
Pragma: no-cache
Cache-Control: no-cache
x-eof: 1
Origin: http://cabat.imod.pasteur.fr:3000
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/52.0.2743.82 Safari/537.36
x-mtok: Q72tzjS9zbmXQW4ih
content-type: text/plain
x-fileId: css36DL8hyPp7sybn
Accept: */*
DNT: 1
Referer: http://cabat.imod.pasteur.fr:3000/widgetboard
Accept-Encoding: gzip, deflate
Accept-Language: en,zh;q=0.8,zh-CN;q=0.6,zh-TW;q=0.4
Cookie: __utma=219899544.213272128.1414491796.1445026431.1445347606.8; EMF_VISIT__MD_414=C2F04569-EC57-4774-A903-91EC1815; _ga=GA1.2.213272128.1414491796; __unam=4127071-1495649a3ce-65939ba9-116; meteor_login_token=eiWZxP_ln4c_--4S3VkqWgYLrPoygncjEsbt79YOwn4; x_mtok=Q72tzjS9zbmXQW4ih

```

**Comment by dr-dimitru** [comment]
Agree, it should be replaced with `protected`. But it's a bit complicated with resumable uploads, which may never be ended and removed by expiration of `continueUploadTTL`.
Currently only `onAfterUpload` was made to be called once

**Comment by oeway** [comment]
Can we add some kind of function such as `onInitiateUpload` which will only execute once before upload?

**Comment by dr-dimitru** [comment]
Yes, could you create separate feature request

**Comment by dr-dimitru** [comment]
About your issue, everything looks good to me.
I'll double check this issue, but currently can't reproduce.
To make me super-sure about this, what next command displays in console?:
```js
Meteor.connection._lastSessionId
```

**Comment by dr-dimitru** [comment]
And, was it working before (*I know this is not first time you're working with MF*)? 
What have you changed? MF version?

**Comment by oeway** [comment]
Before I never really tried to do actually validation, so it's the first time I use this.userId in a callback.

**Comment by oeway** [comment]
```
Meteor.connection._lastSessionId
"BLTK3dLxY6s73gRdW"
```

**Comment by dr-dimitru** [comment]
Okay, give me time to investigate it. 
But this was totally working on latest release, and this feature used by me and other devs
Could you please print all available data in this hook:
```js
console.log(this, arguments);
```

**Comment by oeway** [comment]
```
I20160830-13:31:57.177(2)? { file: 
I20160830-13:31:57.179(2)?    { size: 31995,
I20160830-13:31:57.179(2)?      type: 'image/jpeg',
I20160830-13:31:57.179(2)?      name: 'test.jpeg',
I20160830-13:31:57.179(2)?      meta: 
I20160830-13:31:57.180(2)?       { taskId: 'JxzeySeT5b4NX84Nx',
I20160830-13:31:57.181(2)?         widgetId: 'EejuZZ7munH67mJFt',
I20160830-13:31:57.181(2)?         tags: [] },
I20160830-13:31:57.181(2)?      ext: 'jpeg',
I20160830-13:31:57.181(2)?      extension: 'jpeg',
I20160830-13:31:57.182(2)?      extensionWithDot: '.jpeg',
I20160830-13:31:57.182(2)?      mime: 'image/jpeg',
I20160830-13:31:57.182(2)?      'mime-type': 'image/jpeg',
I20160830-13:31:57.182(2)?      path: 'assets/app/uploads/files/FAemvNTNKWkPQBCYx.jpeg' },
I20160830-13:31:57.182(2)?   chunkId: -1,
I20160830-13:31:57.183(2)?   userId: undefined,
I20160830-13:31:57.183(2)?   user: [Function: user],
I20160830-13:31:57.183(2)?   eof: false } { '0': 
I20160830-13:31:57.183(2)?    { name: 'test.jpeg',
I20160830-13:31:57.184(2)?      extension: 'jpeg',
I20160830-13:31:57.184(2)?      path: 'assets/app/uploads/files/FAemvNTNKWkPQBCYx.jpeg',
I20160830-13:31:57.184(2)?      meta: 
I20160830-13:31:57.185(2)?       { taskId: 'JxzeySeT5b4NX84Nx',
I20160830-13:31:57.185(2)?         widgetId: 'EejuZZ7munH67mJFt',
I20160830-13:31:57.185(2)?         tags: [] },
I20160830-13:31:57.185(2)?      type: 'image/jpeg',
I20160830-13:31:57.186(2)?      size: 31995,
I20160830-13:31:57.186(2)?      versions: { original: [Object] },
I20160830-13:31:57.186(2)?      isVideo: false,
I20160830-13:31:57.186(2)?      isAudio: false,
I20160830-13:31:57.187(2)?      isImage: true,
I20160830-13:31:57.187(2)?      isText: false,
I20160830-13:31:57.187(2)?      isJSON: false,
I20160830-13:31:57.187(2)?      isPDF: false,
I20160830-13:31:57.187(2)?      _storagePath: 'assets/app/uploads/files',
I20160830-13:31:57.188(2)?      _downloadRoute: '/cdn/storage',
I20160830-13:31:57.188(2)?      _collectionName: 'files',
I20160830-13:31:57.188(2)?      _id: 'FAemvNTNKWkPQBCYx' } }
```

**Comment by dr-dimitru** [comment]
Okay, got it, I meant - bug is confirmed

**Comment by dr-dimitru** [comment]
Wait for release. Few hours

**Comment by oeway** [comment]
@dr-dimitru Thanks very much.

**Comment by dr-dimitru** [comment]
@oeway fixed, please update to v1.7.2
Please, let me know how it works on your end

**Comment by oeway** [comment]
@dr-dimitru It worked for me, thanks.
